src/mlpack/methods/ann/brnn_impl.hpp

```diff
@@ -0,0 +1,675 @@
+  }
+  else
+  {
+    results = arma::zeros<arma::cube>(outputSize, batchSize, rho);
+  }
+
+  double performance = 0;
+  size_t responseSeq = 0;
+  arma::mat input;
+
+  reverse(results1.begin(), results1.end());
+  // Performance calculation here.
+  for (size_t seqNum = 0; seqNum < rho; ++seqNum)
+  {
+    if (!single)
+    {
+      responseSeq = seqNum;
+    }
+    boost::apply_visitor(LoadOutputParameterVisitor(
+          std::move(results1)), forwardRNN.network.back());
+    boost::apply_visitor(LoadOutputParameterVisitor(
+          std::move(results2)), backwardRNN.network.back());
+    boost::apply_visitor(ForwardVisitor(std::move(input),
+        std::move(boost::apply_visitor(outputParameterVisitor, mergeLayer))),
+        mergeLayer);
+    boost::apply_visitor(ForwardVisitor(
+        std::move(boost::apply_visitor(outputParameterVisitor, mergeLayer)),
+        std::move(results.slice(seqNum))), mergeOutput);
+    performance += outputLayer.Forward(std::move(results.slice(seqNum)),
+        std::move(arma::mat(responses.slice(responseSeq).colptr(begin),
+        responses.n_rows, batchSize, false, true)));
+  }
+
+  boost::apply_visitor(RunSetVisitor(true), mergeLayer);
+  // Calculate and storing delta parameters from output for t = 1 to T.
+  arma::mat delta;
+  std::vector<arma::mat> allDelta;
+
+  for (size_t seqNum = 0; seqNum < rho; ++seqNum)
+  {
+    if (single && seqNum > 0)
+    {
+      error.zeros();
+    }
+    else if (single && seqNum == 0)
+    {
+      outputLayer.Backward(std::move(results.slice(seqNum)),
+          std::move(arma::mat(responses.slice(0).colptr(begin),
+          responses.n_rows, batchSize, false, true)), std::move(error));
+    }
+    else
+    {
+      outputLayer.Backward(std::move(results.slice(seqNum)),
+          std::move(arma::mat(responses.slice(seqNum).colptr(begin),
+          responses.n_rows, batchSize, false, true)), std::move(error));
+    }
+
+    boost::apply_visitor(BackwardVisitor(std::move(results.slice(seqNum)),
+        std::move(error), std::move(delta)), mergeOutput);
+    allDelta.push_back(arma::mat(delta));
+  }
+
+  // BPTT ForwardRNN from t = T to 1.
+  totalGradient = arma::mat(gradient.memptr(),
+      parameter.n_elem/2, 1, false, false);
+
+  forwardGradient.zeros();
+  forwardRNN.ResetGradients(forwardGradient);
+  backwardGradient.zeros();
+  backwardRNN.ResetGradients(backwardGradient);
+
+  for (size_t seqNum = 0; seqNum < rho; ++seqNum)
+  {
+    forwardGradient.zeros();
+    for (size_t l = 0; l < networkSize; ++l)
+    {
+      boost::apply_visitor(LoadOutputParameterVisitor(
+          std::move(forwardRNNOutputParameter)),
+          forwardRNN.network[networkSize - 1 - l]);
+    }
+    boost::apply_visitor(BackwardVisitor(std::move(
+        boost::apply_visitor(outputParameterVisitor, forwardRNN.network.back())),
+        std::move(allDelta[rho - seqNum - 1]), std::move(delta)), mergeLayer);
```

Comment:
I call `Backward()` here on `mergeLayer` passing input from `forwardRNN` and similarly I have to do this while passing the input from `backwardRNN` (because mergeLayer can only take 1 input). This would call backward for each layer inside mergeLayer with the given input. 
Do you think we could add a parameter to `Backward()` inside mergeLayer so that it calls `Backward()` on a specific layer inside the merge layers? @zoq

Reply:
Or provide another ``Backward`` overload that takes another input?

Reply:
Yes I have added the overload. Please let me know what you think.

Reply:
Looks good to me 👍 ; will take another look at the brrn class tomorrow.

Reply:
okay thanks!